app-typescript/components/WithPagination.tsx

```diff
@@ -0,0 +1,227 @@
+    items: Array<T> | null;
+}
+
+function getPagination(currentPage: number, totalPages: number): Array<number | 'dots'> {
+    let basePages: ReturnType<typeof getPagination> = [
+        currentPage - 2,
+        currentPage - 1,
+        currentPage,
+        currentPage + 1,
+        currentPage + 2,
+    ].filter((page) => page >= 1 && page <= totalPages);
+
+    if (!basePages.includes(1)) { // include first and maybe dots
+        const firstInCurrentList = basePages[0];
+
+        if (firstInCurrentList !== 1) {
+            basePages = [
+                'dots',
+                ...basePages,
+            ];
+        }
+
+        basePages = [
+            1,
+            ...basePages,
+        ];
+    }
+
+    if (!basePages.includes(totalPages)) { // include last and maybe dots
+        const lastInCurrentList = basePages[basePages.length - 1];
+
+        if (lastInCurrentList !== totalPages - 1) { // add dots if we're skipping some numbers
+            basePages = basePages.concat('dots');
+        }
+
+        basePages = [
+            ...basePages,
+            totalPages,
+        ];
+    }
+
+    return basePages;
+}
+
+function getScrollParent(element: HTMLElement | null): HTMLElement | null {
+    if (element == null) {
+        return null;
+    }
+
+    let parentElement: HTMLElement | null = element;
+
+    const overflowY = window.getComputedStyle(parentElement).overflowY;
+    const hasScrollbar = overflowY === 'auto' || overflowY === 'scroll';
+
+    while (parentElement !== null && !hasScrollbar) {
+        parentElement = parentElement.parentElement ?? null;
+    }
+
+    return parentElement;
+}
+
+export class WithPagination<T> extends React.PureComponent<IProps<T>, IState<T>> {
+    private pageCount: number;
+    private abortController: AbortController;
+    private ref: HTMLDivElement | null;
+    private inProgress: boolean;
+
+    constructor(props: IProps<T>) {
+        super(props);
+
+        this.state = {
+            currentPage: 1,
+            items: null,
+        };
+
+        this.switchPage = this.switchPage.bind(this);
+        this.getPageSize = this.getPageSize.bind(this);
+
+        this.pageCount = 0;
+        this.abortController = new window.AbortController(); // window. needed for unit tests
+        this.ref = null;
+        this.inProgress = false;
+    }
+
+    getPageSize() {
+        return this.props.pageSize ?? 20;
+    }
+
+    switchPage(page: number) {
+        if (this.inProgress) {
+            this.abortController.abort();
+        }
+
+        this.inProgress = true;
+        this.props.getItems(page, this.abortController.signal).then((res) => {
+            this.inProgress = false;
+            this.setState({items: res.items, currentPage: page}, () => {
+                const scrollableEl = getScrollParent(this.ref);
+                const diff = scrollableEl != null && this.ref?.scrollHeight != null
+                    ? scrollableEl.offsetHeight - this.ref?.scrollHeight
```

Comment:
> scrollableEl.offsetHeight - this.ref?.scrollHeight

This might evaluate to `200 - undefined` - and we should make sure it doesn't happen. Actually, you do make sure, by only executing it if `this.ref?.scrollHeight != null`. But then you shouldn't be using optional chaining.